### PR TITLE
ENH: Update libexpat files

### PR DESCRIPTION
**Pull request by agravgaard**
The hash method in the old expat would have integer overflows which are
UB. This may have been intentional and not caused error because the UB
result has either been the same on all tested systems or not been relied
upon by anyone.

Related issue in RTK which inspired this PR: https://github.com/SimonRit/RTK/issues/430

## PR Checklist
- [ ] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [ ] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)
- [ ] Added test (or behavior not changed)
- [ ] Updated API documentation (or API not changed)
- [ ] Added [license](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Utilities/KWStyle/ITKHeader.h) to new files (if any)
- [ ] Added Python wrapping to new files (if any) as described in [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) Section 9.5
- [ ] Added [ITK examples](https://github.com/InsightSoftwareConsortium/ITKSphinxExamples) for all new major features (if any)

Refer to the [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) for
further development details if necessary.

### CODE REVIEW

**Comment by agravgaard** [comment]
I don't currently have a Windows machine to test on.
Could anyone give me a clue on how to correct this one:
https://dev.azure.com/itkrobotwindow/ITK.Windows/_build/results?buildId=6576&view=logs&j=2d2b3007-3c5c-5840-9bb0-2b1ea49925f3&t=77aad734-2057-5694-9ae2-ee1f5f26eae8&l=5048

**Comment by SimonRit** [comment]
Have you regenerated `itk_expat_mangle.h` as indicated [here](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/ThirdParty/Expat/src/expat/itk_expat_mangle.h#L30)? It might be something to do...

**Comment by dzenanz** [comment]
Could you break up this PR into more commits? Some preparatory commit(s), the main commit which copies new external code over the existing one, then some follow-up commit(s). That way we don't need to review the third-party code, only the changes in our code, or our changes to the third-party code. You can find recent examples: #2683 and #2708.

File: Modules/ThirdParty/Expat/src/expat/itk_expat_mangle.h [hunk 1]
```diff
@@ -27,79 +27,96 @@ ITK.
...
-#define XML_SetStartNamespaceDeclHandler itk_expat_XML_SetStartNamespaceDeclHandler
-#define XML_SetUnknownEncodingHandler itk_expat_XML_SetUnknownEncodingHandler
-#define XML_SetUnparsedEntityDeclHandler itk_expat_XML_SetUnparsedEntityDeclHandler
-#define XML_SetUserData itk_expat_XML_SetUserData
-#define XML_SetXmlDeclHandler itk_expat_XML_SetXmlDeclHandler
-#define XmlSizeOfUnknownEncoding itk_expat_XmlSizeOfUnknownEncoding
-#define XML_UseParserAsHandlerArg itk_expat_XML_UseParserAsHandlerArg
-#define XmlUtf16Encode itk_expat_XmlUtf16Encode
-#define XmlUtf8Encode itk_expat_XmlUtf8Encode
+#define _XML_DefaultCurrent itk_expat__XML_DefaultCurrent
```
Comment (dzenanz) [inline Modules/ThirdParty/Expat/src/expat/itk_expat_mangle.h:34]: All symbols now having underscore prefix? That seems suspicious. See whether removing them fixes the Windows build failure. Example:
```suggestion
#define XML_DefaultCurrent itk_expat__XML_DefaultCurrent
```

File: Modules/ThirdParty/Expat/src/expat/itk_expat_mangle.h [hunk 2]
```diff
@@ -31,92 +31,92 @@ nm libITKEXPAT-5.3.a | grep " [TRD] " |  awk '{ print $3 }' | sort | awk '{ prin
...
-#define XmlPrologStateInit itk_expat__XmlPrologStateInit
-#define XmlPrologStateInitExternalEntity itk_expat__XmlPrologStateInitExternalEntity
-#define XmlSizeOfUnknownEncoding itk_expat__XmlSizeOfUnknownEncoding
-#define XmlUtf16Encode itk_expat__XmlUtf16Encode
-#define XmlUtf8Encode itk_expat__XmlUtf8Encode
-#define _INTERNAL_trim_to_complete_utf8_characters itk_expat___INTERNAL_trim_to_complete_utf8_characters
-#define testingAccountingGetCountBytesDirect itk_expat__testingAccountingGetCountBytesDirect
-#define testingAccountingGetCountBytesIndirect itk_expat__testingAccountingGetCountBytesIndirect
-#define unsignedCharToPrintable itk_expat__unsignedCharToPrintable
+#define XML_DefaultCurrent itk_expat_XML_DefaultCurrent
```
Comment (dzenanz) [inline Modules/ThirdParty/Expat/src/expat/itk_expat_mangle.h:34]: This commit should be squashed with the previous one.

File: Modules/ThirdParty/Expat/src/expat/expat.h [hunk 3]
```diff
@@ -1,34 +1,133 @@
...
+#ifndef Expat_INCLUDED
+#define Expat_INCLUDED 1
 
 #include <stdlib.h>
-
-#include "expatDllConfig.h"
-
-#if defined(_WIN32) && !defined(ITK_EXPAT_STATIC)
-#  if defined(ITKEXPAT_EXPORTS)
-#    define XMLPARSEAPI(type) __declspec( dllexport ) type __cdecl
```
Comment (dzenanz) [inline Modules/ThirdParty/Expat/src/expat/expat.h:15]: Where did this go? Disappearance of this is the most likely cause for [link failures](https://open.cdash.org/viewBuildError.php?buildid=7479711) on Windows.

Comment (agravgaard) [reply Modules/ThirdParty/Expat/src/expat/expat.h]: Nice catch!

File: Modules/ThirdParty/Expat/src/expat/itk_expat_mangle.h [hunk 4]
```diff
@@ -97,9 +96,27 @@ nm libITKEXPAT-4.9.a | grep " [TRD] " |  awk '{ print $3 }' | sort | awk '{ prin
 #define XML_SetUnparsedEntityDeclHandler itk_expat_XML_SetUnparsedEntityDeclHandler
 #define XML_SetUserData itk_expat_XML_SetUserData
 #define XML_SetXmlDeclHandler itk_expat_XML_SetXmlDeclHandler
-#define XmlSizeOfUnknownEncoding itk_expat_XmlSizeOfUnknownEncoding
+#define XML_StopParser itk_expat_XML_StopParser
+#define XML_UseForeignDTD itk_expat_XML_UseForeignDTD
 #define XML_UseParserAsHandlerArg itk_expat_XML_UseParserAsHandlerArg
+#define XmlGetUtf16InternalEncoding itk_expat_XmlGetUtf16InternalEncoding
```
Comment (dzenanz) [inline Modules/ThirdParty/Expat/src/expat/itk_expat_mangle.h:102]: This got re-sorted here from line 50. Does [`export LC_COLLATE=POSIX`](https://github.com/InsightSoftwareConsortium/ITK/blob/v5.3rc01/Modules/ThirdParty/HDF5/src/itkhdf5/src/itk_hdf5_mangle.h#L17) give the same sort order as before? Or maybe [some other common choice](https://stackoverflow.com/questions/23660247/treatment-of-spaces-in-sort-command-difference-between-lc-collate-c-and-lc-coll)?

File: Modules/ThirdParty/Expat/src/expat/itk_expat_mangle.h [hunk 5]
```diff
@@ -27,46 +27,43 @@ ITK.
 
 The following command was used to obtain the symbol list:
 
-nm libITKEXPAT-4.9.a | grep " [TRD] " |  awk '{ print $3 }' | sort | awk '{ print "#define "$1" itk_expat_"$1 }'
+nm libITKEXPAT-5.3.a | grep " [TRD] " |  awk '{ print $3 }' | sort | awk '{ print "#define "$1" itk_expat_"$1 }'
```
Comment (dzenanz) [inline Modules/ThirdParty/Expat/src/expat/itk_expat_mangle.h:30]: Putting the used LC_COLLATE here would be helpful for future updates.

**Review by dzenanz — APPROVED** [review]
This is good even as-is. The remaining remarks are optional.

File: Modules/ThirdParty/Expat/src/expat/itk_expat_mangle.h (hunk 4 above)
Comment (agravgaard) [reply Modules/ThirdParty/Expat/src/expat/itk_expat_mangle.h]: Changing `LC_COLLATE` didn't change the output for me, but using `-d` and `-f` flags for sort seemed to do the job.
```
  -d, --dictionary-order      consider only blanks and alphanumeric characters
  -f, --ignore-case           fold lower case to upper case characters
```

**Review by dzenanz — APPROVED** [review]
Awesome!

**Comment by dzenanz** [comment]
To be merged when CI comes back green.

**Pull request merged**